src/lib.rs

```diff
@@ -881,6 +954,96 @@ where
 }
 
 impl Buffer {
+    /// Map the buffer for reading. The result is returned in a future.
+    /// The caller is responsible for unmapping the buffer once the future is completed.
+    pub fn map_read<T>(&self, start: BufferAddress, size: BufferAddress) -> GpuFuture<BufferMapReadResult<T>>
+    where
+        T: 'static + FromBytes,
+    {
+        let (future, completion) = future::new_gpu_future(self.device_id);
+
+        extern "C" fn buffer_map_read_future_wrapper<T>(
```

Comment:
can this function use `T` type from the parent function?

Reply:
No, see `rustc --explain E0401`.